sgkit/tests/test_association.py

```diff
@@ -0,0 +1,169 @@
+import warnings
+from typing import Any, Dict, List, Optional, Tuple
+
+import numpy as np
+import pandas as pd
+import pytest
+import xarray as xr
+from pandas import DataFrame
+from xarray import Dataset
+
+from sgkit.stats.association import gwas_linear_regression
+from sgkit.typing import ArrayLike
+
+with warnings.catch_warnings():
+    warnings.simplefilter("ignore", DeprecationWarning)
+    # Ignore: DeprecationWarning: Using or importing the ABCs from 'collections'
+    # instead of from 'collections.abc' is deprecated since Python 3.3,
+    # and in 3.9 it will stop working
+    import statsmodels.api as sm
+    from statsmodels.regression.linear_model import RegressionResultsWrapper
+
+
+def _generate_test_data(
+    n: int = 100,
+    m: int = 10,
+    p: int = 3,
+    e_std: float = 0.001,
+    b_zero_slice: Optional[slice] = None,
+    seed: Optional[int] = 1,
+) -> Tuple[ArrayLike, ArrayLike, ArrayLike, ArrayLike]:
+    """Test data simulator for multiple variant associations to a continuous outcome
+
+    Outcomes for each variant are simulated separately based on linear combinations
+    of randomly generated fixed effect covariates as well as the variant itself.
+
+    This does not add an intercept term in covariates.
+
+    Parameters
+    ----------
+    n : int, optional
+        Number of samples
+    m : int, optional
+        Number of variants
+    p : int, optional
+        Number of covariates
+    e_std : float, optional
+        Standard deviation for noise term
+    b_zero_slice : slice
+        Variant beta values to zero out, defaults to `slice(m // 2)`
+        meaning that the first half will all be 0.
+        Set to `slice(0)` to disable.
+
+    Returns
+    -------
+    g : (n, m) array-like
+        Simulated genotype dosage
+    x : (n, p) array-like
+        Simulated covariates
+    bg : (m,) array-like
+        Variant betas
+    ys : (m, n) array-like
+        Outcomes for each column in genotypes i.e. variant
+    """
+    if b_zero_slice is None:
+        b_zero_slice = slice(m // 2)
+    rs = np.random.RandomState(seed)
+    g = rs.uniform(size=(n, m), low=0, high=2)
+    x = rs.normal(size=(n, p))
+    bg = rs.normal(size=m)
+    bg[b_zero_slice or slice(m // 2)] = 0
+    bx = rs.normal(size=p)
+    e = rs.normal(size=n, scale=e_std)
+
+    # Simulate y values using each variant independently
+    ys = np.array([g[:, i] * bg[i] + x @ bx + e for i in range(m)])
+    return g, x, bg, ys
+
+
+def _generate_test_dataset(**kwargs: Any) -> Dataset:
+    g, x, bg, ys = _generate_test_data(**kwargs)
+    data_vars = {}
+    data_vars["dosage"] = (["variant", "sample"], g.T)
+    for i in range(x.shape[1]):
+        data_vars[f"covar_{i}"] = (["sample"], x[:, i])
+    for i in range(len(ys)):
+        data_vars[f"trait_{i}"] = (["sample"], ys[i])
+    attrs = dict(beta=bg)
+    return xr.Dataset(data_vars, attrs=attrs)  # type: ignore[arg-type]
+
+
+@pytest.fixture  # type: ignore[misc]
```

Comment:
Is this fixture intentionally with function scope?

Reply:
Nah thanks!  Will change to module scope

Reply:
https://github.com/pystatgen/sgkit/pull/16/commits/273375e5d949ae3cd6b22572a6fccb2bd3064f0f